source/common/upstream/upstream_impl.h

```diff
@@ -387,6 +388,9 @@ class HostSetImpl : public HostSet {
       HostsPerLocalityConstSharedPtr all_hosts_per_locality,
       LocalityWeightsConstSharedPtr locality_weights, uint32_t overprovisioning_factor);
 
+  static absl::optional<uint32_t>
+  chooseLocality(std::unique_ptr<EdfScheduler<LocalityEntry>>& locality_scheduler);
```

Comment:
nit: this should probably just take an `EdfScheduler<LocalityEntry>&`

Reply:
Since it checks for nullptr I can't do that without adding null checks elsewhere, but I guess I could change it to raw pointer

Reply:
That'd work, or no change would be fine too :)

Reply:
Passing a raw pointer in this case is more common, so I would probably do that, though it doesn't matter that much.